vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+                     size_t(edge), size_t(edge_src_node(edge)));
+        all_valid = false;
+    }
+    return all_valid;
+}
+
+/* Check if source nodes of a edge are all valid */
+bool RRGraph::check_edge_sink_nodes() const {
+    bool all_valid = true;
+    for (auto edge : edges()) {
+        if (true == check_edge_sink_node(edge)) {
+            continue;
+        }
+        /* Reach here, it means there is something wrong! 
+         * Print a warning  
+         */
+        VTR_LOG_WARN("Edge %d has a invalid sink node %d!\n",
+                     size_t(edge), size_t(edge_sink_node(edge)));
+        all_valid = false;
+    }
+    return all_valid;
+}
+
+/* This function aims at checking any duplicated edges (with same EdgeId) 
+ * of a given node. 
+ * We will walkthrough the input edges of a node and see if there is any duplication
+ */
+bool RRGraph::check_node_duplicated_edges(RRNodeId node) const {
+    bool no_duplication = true;
+    /* Check each input edges */
+    for (size_t iedge = 0; iedge < node_in_edges_[node].size(); ++iedge) {
+        for (size_t jedge = iedge + 1; jedge < node_in_edges_[node].size(); ++jedge) {
+            if (node_in_edges_[node][iedge] != node_in_edges_[node][jedge]) {
+                /* Normal case, continue */
+                continue;
+            }
+            /* Reach here it means we find some duplicated edges and report errors */
+            /* Print a warning! */
+            VTR_LOG_WARN("Node %d has duplicated input edges (edgeId:%d and %d)!\n",
+                         size_t(node), iedge, jedge);
+            this->print_node(node);
+            no_duplication = false;
+        }
+    }
+
+    return no_duplication;
+}
+
+/* Check the whole Routing Resource Graph  
+ * identify and report any duplicated edges between two nodes 
+ */
+bool RRGraph::check_duplicated_edges() const {
+    bool no_duplication = true;
+    /* For each node:
+     * Search input edges, see there are two edges with same id or address 
+     */
+    for (auto node : nodes()) {
+        if (false == check_node_duplicated_edges(node)) {
+            no_duplication = false;
+        }
+    }
+
+    return no_duplication;
+}
+
+/* 
+ * identify and report any duplicated edges between two nodes 
+ */
+bool RRGraph::check_dangling_nodes() const {
+    bool no_dangling = true;
+    /* For each node: 
+     * check if the number of input edges and output edges are both 0
+     * If so, this is a dangling nodes and report 
+     */
+    for (auto node : nodes()) {
+        if ((0 == this->node_fan_in(node))
+            && (0 == this->node_fan_out(node))) {
+            /* Print a warning! */
+            VTR_LOG_WARN("Node %s is dangling (zero fan-in and zero fan-out)!\n",
+                         node);
+            VTR_LOG_WARN("Node details for debugging:\n");
+            this->print_node(node);
+            no_dangling = false;
+        }
+    }
+
```

Comment:
Another related but useful check would be to verify that:
* Only SOURCES have zero fanin and non-zero fanout
* Only SINKS have non-zero fanin and zero fanout

Reply:
Done